frontend/model/captureLogs.js

```diff
@@ -0,0 +1,250 @@
+  }
+}
+
+function verifyLogsSize () {
+  let lastEntriesCount = null
+  let failedDeletionCount = 0
+  let markers = null
+
+  // Delete nth oldest logs recursively. This scenario can happen when the
+  // entries limit is changed. Example: There are 400 logs and the limit is 500.
+  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+  // We do it recursively in chunks of 25 until there's only 75 logs again.
+  while (entriesCount >= ENTRIES_LIMIT && failedDeletionCount < 3) {
+    markers = markers || getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+
+    if (entriesCount === lastEntriesCount) { failedDeletionCount++ }
+    lastEntriesCount = entriesCount
+
+    do {
+      const log = JSON.parse(giLSget(oldestEntry)) || {}
+      giLSremove(oldestEntry)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+  }
+
+  if (failedDeletionCount < 3) {
+    giLSset('markers', JSON.stringify(markers))
+    giLSset('count', entriesCount)
+  } else {
+    // - There are too many entriesCount, however, for some unknown reason
+    //   we weren't able to remove them.
+    // - Capture a new log error manually (withou console.error) so
+    //   that verifyLogsSize() doesn't get called again. There's no need
+    //   for that and it would cause a loop of failed deletions.
+    captureLog('error', 'verifyLogsSize() failed! Unable to delete oldest logs.')
+
+    // To avoid infinite loops or memory leaks, let's just pause the capturing.
+    captureLogsPause()
+    // Or... if we want to be extreme, clear the logs and have a fresh start?
+    // clearLogs()
+  }
+}
+
+function verifyLogsConfigs () {
+  // If ENTRIES_LIMIT or ENTRIES_MARKER_NTH are changed in a release
+  // we need to recalculate markers and verify if it reached the size limit.
+  const storedLimit = +giLSget('limit')
+  const storedMarkerNth = +giLSget('markerNth')
+
+  // when it's the first time running this, just set the correct limits
+  if (storedLimit === 0) {
+    giLSset('limit', ENTRIES_LIMIT)
+    giLSset('markerNth', ENTRIES_MARKER_NTH)
+    return
+  }
+
+  if (storedMarkerNth !== ENTRIES_MARKER_NTH) {
+    // recalculate markers based on new nth rule.
+    const newMarkers = []
+    let entryIndex = entriesCount
+    let prevEntry = lastEntry
+    do {
+      const log = JSON.parse(giLSget(`${prevEntry}`))
+      if (!log) break
+      if (entryIndex % ENTRIES_MARKER_NTH === 0) {
+        newMarkers.push(prevEntry)
+      }
+      entryIndex--
+      prevEntry = log.prev
+    } while (prevEntry)
+    // reverse new markers to be in chronological order, and then store them.
+    giLSset('markers', JSON.stringify(newMarkers.reverse()))
+    giLSset('markerNth', ENTRIES_MARKER_NTH)
+  }
+
+  if (storedLimit !== ENTRIES_LIMIT) {
+    giLSset('limit', ENTRIES_LIMIT)
+    verifyLogsSize()
+  }
+}
```

Comment:
I was thinking of how to make it so that `clearLogs()` can reset everything and came up with the code below, do you think it can be adopted? See my other comments in `clearLogs()` and `captureLogsStart()`.

```js
function verifyLogsConfigs () {
  lastEntry = giLSget('lastEntry')
  // calculate entriesCount to ensure we have accurate number
  entriesCount = 0
  let curEntry = lastEntry
  while (curEntry) {
    curEntry = curEntry.prev
    ++entriesCount
  }

  // If ENTRIES_LIMIT or ENTRIES_MARKER_NTH are changed in a release
  // we need to recalculate markers and verify if it reached the size limit.
  const storedLimit = +giLSget('limit')
  const storedMarkerNth = +giLSget('markerNth')

  giLSset('count', entriesCount)
  giLSset('limit', ENTRIES_LIMIT)
  giLSset('markerNth', ENTRIES_MARKER_NTH)

  if (storedMarkerNth !== ENTRIES_MARKER_NTH) {
    // recalculate markers based on new nth rule.
    const newMarkers = []
    let entryIndex = entriesCount
    curEntry = lastEntry
    while (curEntry) {
      const log = JSON.parse(giLSget(curEntry))
      if (!log) break
      if (entryIndex % ENTRIES_MARKER_NTH === 0) {
        newMarkers.push(prevEntry)
      }
      entryIndex--
      curEntry = log.prev
    }
    // reverse new markers to be in chronological order, and then store them.
    giLSset('markers', JSON.stringify(newMarkers.reverse()))
  }
  // delete old logs if the limit was reduced
  if (storedLimit > ENTRIES_LIMIT) {
    verifyLogsSize()
  }
}
```

Reply:
Did you run this code?... So you are suggesting that we verify the `entriesCount` on `captureLogsStart` and `verifyLogsConfigs`. Doing the same thing twice when I think we shouldn't do it even once... (see previous comment explaining why).

Then here you are proposing to change `limit` and `markerNth` every time, when it's only needed the first time is run and when those configs are changed. If we want to care about performance, we should only hit LS when absolutely necessary.

And last but not least, your approach calls `verifyLogsSize()` always. Once again, we only need to call it if the limit changed. Then it will be called after capturing a new log only if the entries count reached the limit.

Reply:
Here's it's ok because `verifyLogsConfigs()` isn't on the hot path (i.e. it's not called on each log).

Reply:
> Here you are proposing to set limit and markerNth every time

And yes, I'm suggesting that because it makes the code much easier to think about and less likely to break.

`verifyLogsConfigs()` needs to be called under multiple different circumstances. But not having unnecessary `if` branches we ensure this function gets a clean state each time it's called, for example from `clearLogs()` when there's an error in `verifyLogSize()`.

Reply:
Question: when we launch a new release with new limits, do we ask the user to log in again?

If yes, we don't need this `verifyLogsConfig()` at all because all the logs were wiped out on logout, so we don't need to recount entries or recalculate markers. We can just remove this function. and call `clearLogs()` instead to make sure we have a fresh start.

What do you think?